### PR TITLE
Feature/5972 index algorithm optimization

### PR DESCRIPTION
**Pull request by HavierD**
Features are not perfect yet. It will possibly be optimized in the future according to the frontend performance tests.

### CODE REVIEW

File: indexer/src/main/java/au/org/aodn/esindexer/service/IndexerServiceImpl.java [hunk 1]
```diff
@@ -257,33 +261,45 @@ public CompletableFuture<ResponseEntity<String>> indexMetadata(String metadataVa
     }
 
     @Override
-    public CompletableFuture<ResponseEntity<String>> indexDataset(Dataset dataset) {
-        try {
+    public List<BulkResponse> indexDataset(String uuid, LocalDate startDate, LocalDate endDate) {
```
Comment (utas-raymondng) [inline indexer/src/main/java/au/org/aodn/esindexer/service/IndexerServiceImpl.java:264]: Without using ComplableFuture, you may get timeout for long running operation, you think it is ok?

**Review by utas-raymondng — CHANGES_REQUESTED** [review]

**Comment by HavierD** [comment]
https://github.com/aodn/backlog/issues/6081      new ticket to do the refactoring

File: indexer/src/main/java/au/org/aodn/esindexer/service/DataAccessServiceImpl.java [hunk 2]
```diff
@@ -35,11 +34,11 @@ public DataAccessServiceImpl(String serverUrl) {
 
     private final RestTemplate restTemplate = new RestTemplate();
```
Comment (utas-raymondng) [inline indexer/src/main/java/au/org/aodn/esindexer/service/DataAccessServiceImpl.java:35]: Please user @Autowired for restTemplate

**Review by utas-raymondng — CHANGES_REQUESTED** [review]
one more minor change

**Review by utas-raymondng — APPROVED** [review]
LGTM

**Comment by utas-raymondng** [comment]
> [aodn/backlog#6081](https://github.com/aodn/backlog/issues/6081) new ticket to do the refactoring

I guess you will add test case to cover your change right?

**Comment by HavierD** [comment]
> > [aodn/backlog#6081](https://github.com/aodn/backlog/issues/6081) new ticket to do the refactoring
> 
> I guess you will add test case to cover your change right?

Yes all the new features about real dataset in es-indexer, ogc-api, and frontend will be added test cases when the features are all stable. Currently i am still changing them.

**Pull request merged**